File: java/com/pushtechnology/diffusion/examples/ControlClientChangingSystemAuthentication.java
/*******************************************************************************
 * Copyright (C) 2014, 2015 Push Technology Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *******************************************************************************/
package com.pushtechnology.diffusion.examples;

import java.util.HashSet;
import java.util.Set;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.pushtechnology.diffusion.client.Diffusion;
import com.pushtechnology.diffusion.client.callbacks.ErrorReason;
import com.pushtechnology.diffusion.client.features.control.clients.SystemAuthenticationControl;
import com.pushtechnology.diffusion.client.features.control.clients.SystemAuthenticationControl.ConfigurationCallback;
import com.pushtechnology.diffusion.client.features.control.clients.SystemAuthenticationControl.ScriptBuilder;
import com.pushtechnology.diffusion.client.features.control.clients.SystemAuthenticationControl.SystemAuthenticationConfiguration;
import com.pushtechnology.diffusion.client.features.control.clients.SystemAuthenticationControl.SystemPrincipal;
import com.pushtechnology.diffusion.client.features.control.clients.SecurityStoreFeature.UpdateStoreCallback;
import com.pushtechnology.diffusion.client.session.Session;

/**
 * An example of using a control client to alter the system authentication
 * configuration.
 * <P>
 * This uses the {@link SystemAuthenticationControl} feature only.
 *
 * @author Push Technology Limited
 * @since 5.2
 */
public class ControlClientChangingSystemAuthentication {

    private static final Logger LOG =
        LoggerFactory.getLogger(
            ControlClientChangingSystemAuthentication.class);

    private final SystemAuthenticationControl systemAuthenticationControl;

    /**
     * Constructor.
     */
    public ControlClientChangingSystemAuthentication() {

        final Session session = Diffusion.sessions()
            // Authenticate with a user that has the VIEW_SECURITY and
            // MODIFY_SECURITY permissions.
            .principal("admin").password("password")
            // Use a secure channel because we're transferring sensitive
            // information.
            .open("wss://diffusion.example.com:80");

        systemAuthenticationControl =
            session.feature(SystemAuthenticationControl.class);
    }

    /**
     * For all system users, update the assigned roles to replace the
     * "SUPERUSER" role and with "ADMINISTRATOR".
     *
     * @param callback result callback
     */
    public void changeSuperUsersToAdministrators(UpdateStoreCallback callback) {

        systemAuthenticationControl.getSystemAuthentication(
            new ChangeSuperUsersToAdministrators(callback));
    }

    private final class ChangeSuperUsersToAdministrators
        implements ConfigurationCallback {

        private final UpdateStoreCallback callback;

        ChangeSuperUsersToAdministrators(UpdateStoreCallback callback) {
            this.callback = callback;
        }

        @Override
        public void onReply(SystemAuthenticationConfiguration configuration) {

            ScriptBuilder builder =
                systemAuthenticationControl.scriptBuilder();

            // For all system users ...
            for (SystemPrincipal principal : configuration.getPrincipals()) {

                final Set<String> assignedRoles = principal.getAssignedRoles();

                // ... that have the SUPERUSER assigned role ...
                if (assignedRoles.contains("SUPERUSER")) {
                    final Set<String> newRoles = new HashSet<>(assignedRoles);
                    newRoles.remove("SUPERUSER");
                    newRoles.add("ADMINISTRATOR");

                    // ... add a command to the script that updates the user's
                    // assigned roles, replacing SUPERUSER with "ADMINISTRATOR".
                    builder =
                        builder.assignRoles(principal.getName(), newRoles);
                }
            }

            final String script = builder.script();

            LOG.info(
                "Sending the following script to the server:\n{}",
                script);

            systemAuthenticationControl.updateStore(
                script,
                callback);
        }

        @Override
        public void onError(ErrorReason errorReason) {
            // This might fail if the session lacks the required permissions.
            callback.onError(errorReason);
        }
    }

    /**
     * Close the session.
     */
    public void close() {
        systemAuthenticationControl.getSession().close();
    }
}

File: dotnet/examples/ControlClientChangingSystemAuthentication.cs
/**
 * Copyright © 2014, 2015 Push Technology Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Collections.Generic;
using System.Linq;
using PushTechnology.ClientInterface.Client.Callbacks;
using PushTechnology.ClientInterface.Client.Factories;
using PushTechnology.ClientInterface.Client.Features.Control.Clients;
using PushTechnology.ClientInterface.Client.Types;
using IUpdateStoreCallback = PushTechnology.ClientInterface.Client.Features.Control.Clients.SecurityControl.IUpdateStoreCallback;

namespace Examples
{
    public class ControlClientChangingSystemAuthentication
    {
        #region Fields

        private readonly ISystemAuthenticationControl theSystemAuthenticationControl;

        #endregion Fields

        #region Constructor

        public ControlClientChangingSystemAuthentication()
        {
            var session =
                Diffusion.Sessions
                    // Authenticate with a user that has the VIEW_SECURITY and MODIFY_SECURITY permissions
                    .Principal( "control" ).Password( "password" )
                    // Use a secure channel because we're transferring sensitive information.
                    .Open( "dpt://localhost:8080" );

            theSystemAuthenticationControl = session.GetSystemAuthenticationControlFeature();

            session.Start();
        }

        #endregion Constructor

        #region Public Methods

        /// <summary>
        /// For all system users, update the assigned roles to replace the 'SUPERUSER' role and with 'ADMINISTRATOR'.
        /// </summary>
        /// <param name="callback"></param>
        public void ChangeSuperUsersToAdministrators( IUpdateStoreCallback callback )
        {
            theSystemAuthenticationControl.GetSystemAuthentication( 
                new ChangeSuperusersToAdministrators( theSystemAuthenticationControl, callback ) );
        }

        #endregion Public Methods

        #region Private Classes

        private class InternalUpdateStoreCallback : IUpdateStoreCallback
        {
            /// <summary>
            /// The script was applied successfully.
            /// </summary>
            public void OnSuccess()
            {
            }

            /// <summary>
            /// The script was rejected.  No changes were made to the system authentication store.
            /// </summary>
            /// <param name="errors">The details of why the script was rejected.</param>
            public void OnRejected( ICollection<IErrorReport> errors )
            {
            }

            /// <summary>
            /// Notification of a contextual error related to this callback. This is
            /// analogous to an exception being raised. Situations in which
            /// <code>OnError</code> is called include the session being closed, a
            /// communication timeout, or a problem with the provided parameters. No
            /// further calls will be made to this callback.
            /// </summary>
            /// <param name="errorReason">errorReason a value representing the error; this can be one of
            /// constants defined in <see cref="ErrorReason" />, or a feature-specific
            /// reason.</param>
            public void OnError( ErrorReason errorReason )
            {
            }
        }

        private class ChangeSuperusersToAdministrators : IConfigurationCallback
        {
            #region Fields

            private readonly ISystemAuthenticationControl theSystemAuthenticationControl;
            private readonly IUpdateStoreCallback theCallback;

            #endregion Fields

            #region Constructor

            public ChangeSuperusersToAdministrators( 
                ISystemAuthenticationControl systemAuthenticationControl, 
                IUpdateStoreCallback callback )
            {
                theSystemAuthenticationControl = systemAuthenticationControl;
                theCallback = callback;
            }

            #endregion Constructor

            /// <summary>
            /// The configuration callback reply.
            /// </summary>
            /// <param name="systemAuthenticationConfiguration">The system authenticationConfiguration stored on the server.</param>
            public void OnReply( ISystemAuthenticationConfiguration systemAuthenticationConfiguration )
            {
                var builder = theSystemAuthenticationControl.ScriptBuilder();

                // For all system users...
                foreach( var principal in systemAuthenticationConfiguration.Principals )
                {
                    var assignedRoles = principal.AssignedRoles;

                    // ...that have the 'SUPERUSER' assigned role...
                    if( !assignedRoles.Contains( "SUPERUSER" ) ) continue;

                    var newRoles = new HashSet<string>( assignedRoles );

                    newRoles.Remove( "SUPERUSER" );
                    newRoles.Add( "ADMINISTRATOR" );

                    // ...and add a command to the script that updates the user's assigned roles, replacing 'SUPERUSER' with
                    // 'ADMINISTRATOR'.
                    builder = builder.AssignRoles( principal.Name, newRoles.ToList() );
                }

                var script = builder.Script();

                theSystemAuthenticationControl.UpdateStore( script, theCallback );
            }

            /// <summary>
            /// Notification of a contextual error related to this callback. This is
            /// analogous to an exception being raised. Situations in which
            /// <code>OnError</code> is called include the session being closed, a
            /// communication timeout, or a problem with the provided parameters. No
            /// further calls will be made to this callback.
            /// </summary>
            /// <param name="errorReason">errorReason a value representing the error; this can be one of
            /// constants defined in <see cref="ErrorReason" />, or a feature-specific
            /// reason.</param>
            public void OnError( ErrorReason errorReason )
            {
                theCallback.OnError( errorReason );
            }
        }

        #endregion Private Classes

        #region Tests

        //[Test]
        public void ControlClientChangingSystemAuthenticationTest()
        {
            var client = new ControlClientChangingSystemAuthentication();
            
            client.ChangeSuperUsersToAdministrators( new InternalUpdateStoreCallback() );
        }

        #endregion Tests
    }
}